include/execution.hpp

```diff
@@ -1692,6 +1676,111 @@ namespace std::execution {
       }
     } transfer {};
   } // namespace __transfer
+
+  inline namespace __on {
+    namespace __impl {
+      template <class R_>
+        struct __receiver_ref : receiver_adaptor<__receiver_ref<R_>> {
+          using R = __t<R_>;
+          R& __r_;
+          R&& base() && noexcept { return (R&&) __r_; }
+          const R& base() const & noexcept { return __r_; }
+        };
+
+      template <class Sch_, class S_>
+        struct __sender : sender_base {
+          using Sch = __t<Sch_>;
+          using S = __t<S_>;
+
+          Sch __sch_;
+          S __s_;
+
+          template <class R_>
+            struct __operation;
+
+          template <class R_>
+            struct __receiver : receiver_adaptor<__receiver<R_>> {
+              using R = __t<R_>;
+              __operation<R_>* __op_;
```

Comment:
So the inner receiver needs essentially all the state. And we never give it out to the user. 

So what I thought an simpler setup is to store `sch`, `__r_` and `__s_` inside the receiver and make that a member of the operation_state. 

Yes we would construct `r` when `connect` is called and not when `start` is called but that should be opaque. That way  it is much simpler to reason about lifetimes

Reply:
I don't think this works. Once `set_value` is called on the inner receiver, its lifetime effectively ends, which will be true if you decide, like I did, to reuse the storage for the intermediate operation state. You don't want that teardown to destroy _all_ your state; then you can't launch the inner sender.

Also be mindful of the fact that receivers are copied/moved when calling connect. You must take care that the copy/move isn't potentially throwing, or else that if it does throw, you still have a valid receiver on which you can call `set_error`.